runtime/src/ecalls.rs

```diff
@@ -2,24 +2,42 @@ pub use core::fmt::Write;
 
 #[cfg(target_arch = "riscv32")]
 mod riscv32 {
+    extern crate alloc;
+    use serde::{de::DeserializeOwned, Serialize};
+
     // To simplify calling out to the environment, we keep the
     // argument registers intact, and place the function number
     // in s2 (rust will not allow us to use s0 or s1).
 
     macro_rules! ecall {
-    ($n:literal,$inp1:expr,$inp2:expr,$out:expr) => {
-        unsafe {
-            core::arch::asm!("ecall", in("s2") $n, in("a1") $inp1, in("a2") $inp2, out("a0") $out)
+        ($n:literal,$inp1:expr,$inp2:expr,$out:expr) => {
+            unsafe {
+                core::arch::asm!("ecall", in("s2") $n, in("a1") $inp1, in("a2") $inp2, out("a0") $out)
+            }
         }
     }
-}
 
     /// Write a string to the output console (if any).
     pub fn write_log(s: &str) {
         let mut _out: u32;
         ecall!(1, s.as_ptr(), s.len(), _out);
     }
 
+    /// Read an object off the private input tape
+    pub fn read_private_input<T: DeserializeOwned>() -> Result<T, postcard::Error> {
+        let inp: u32 = 0;
+        let mut out: u32 = 0;
+
+        let mut bytes = alloc::vec::Vec::<u8>::new();
+
+        while out != u32::MAX {
+            ecall!(2, inp, inp, out);
+            bytes.push(out.to_le_bytes()[0])
+        }
```

Comment:
nits:
1. Document that it exhausts the private tape, and essentially can only be called once.
2.
```rust
let bytes: Vec<u8> = core::iter::from_fn(read_from_private_input).collect();
postcard::from_bytes::<T>(bytes.as_slice())
```
to have a single implementation wrapping the ecall
3. Let's add native counterparts so that compilation at least doesn't fail.